### ISSUE TITLE
compile.with_all_deps fails with puppet_gem provider

### ISSUE THREAD

**Issue by ProgrammingAce** [rodjek/rspec-puppet#352]
Hey folks, I can't figure out how to get compile.with_all_deps to pass when using the puppet_gem package provider. It fails with:

```
  1) testing supported operating systems on redhat-7-x86_64 testing class without any parameters should compile into a catalogue without dependency cycles
     Failure/Error: it { is_expected.to compile.with_all_deps }
       error during compilation: Parameter provider failed on Package[r10k]: Invalid package provider 'puppet_gem' at /Users/pmcdonough/code/puppet-modules/testing/spec/fixtures/modules/testing/manifests/install.pp:9
     # ./spec/classes/example_spec.rb:14:in `block (6 levels) in <top (required)>'
```

I'm running on OSX version:10.11.2
With puppet version: 4.3.1
With rspec-puppet: 2.3.0

This is my entire test file:

```
require 'spec_helper'

describe 'testing' do
  context 'supported operating systems' do
    on_supported_os.each do |os, facts|
      context "on #{os}" do
        let(:facts) do
          facts
        end
        context "testing class without any parameters" do

          it { is_expected.to compile.with_all_deps }
          it { is_expected.to contain_package('r10k').with_provider("puppet_gem") }

        end
      end
    end
  end
end
```

And this is the only non-boilerplate code in the module:

```
class testing::install {
    package { 'r10k':
        ensure => 'installed',
        provider => 'puppet_gem',
    }
}
```

The module successfully applies to real instances.

**Comment by DavidS** [comment]
Looking at the [docs](https://docs.puppetlabs.com/references/latest/type.html#package-provider-puppet_gem), the puppet_gem provider is only available if there is a local puppet installation found in /opt. rspec-puppet currently has no mechanism to stub that out.

I fear this is related to the changes in #261 where the catalog was expanded using to_ral. /cc @raphink @mcanevet

**Comment by raphink** [comment]
That could indeed be a side effect of to_ral I'm afraid...

**Comment by DavidAwad** [comment]
Are there any fixes to this? Are there any ways to force rspec-puppet to accept stubs for custom providers? I'm having a similar problem where I'm using puppet homebrew, but rspec fails to accept it as a valid package provider and fails the tests.

**Comment by rodjek** [comment]
This should now be resolved by #498
